### PR TITLE
Handle null values in TempData properly

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Issue #5243 

@pranavkm @kichalla

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/SessionStateTempDataProvider.cs [hunk 1]
```diff
@@ -72,76 +74,83 @@ public class SessionStateTempDataProvider : ITempDataProvider
...
+                    var convertedDictionary = new Dictionary<string, object>(
+                        tempDataDictionary,
+                        StringComparer.OrdinalIgnoreCase);
+                    foreach (var item in tempDataDictionary)
                     {
-                        var arrayType = jArrayValue[0].Type;
-                        Type returnType;
-                        if (_tokenTypeLookup.TryGetValue(arrayType, out returnType))
+                        var jArrayValue = item.Value as JArray;
+                        var jObjectValue = item.Value as JObject;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/SessionStateTempDataProvider.cs]: Pattern matching will fix this and earn mucho mucho PranavPoints

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/SessionStateTempDataProvider.cs [hunk 2]
```diff
@@ -72,76 +74,83 @@ public class SessionStateTempDataProvider : ITempDataProvider
                     tempDataDictionary = _jsonSerializer.Deserialize<Dictionary<string, object>>(writer);
                 }
 
-                var convertedDictionary = new Dictionary<string, object>(
-                    tempDataDictionary,
-                    StringComparer.OrdinalIgnoreCase);
-                foreach (var item in tempDataDictionary)
+                if (tempDataDictionary != null)
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/SessionStateTempDataProvider.cs]: Could we just return here if it happens to be null? One fewer indentation.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewFeatures/SessionStateTempDataProviderTest.cs [hunk 3]
```diff
@@ -54,6 +54,21 @@ public void Load_ReturnsEmptyDictionary_WhenNoSessionDataIsAvailable()
             Assert.Empty(tempDataDictionary);
         }
 
+        [Fact]
+        public void Load_ReturnsEmptyDictionary_WhenSessionDataIsSetToNull()
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewFeatures/SessionStateTempDataProviderTest.cs]: `Load_ReturnsEmptyDictionary_WhenSessionDataIsEmpty`? It's not null is it?

Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewFeatures/SessionStateTempDataProviderTest.cs]: 👍 I meant it more like `WhenSessionDataGetsDeserializedToNull`. Will change

**Comment by ajaybhargavb** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/SessionStateTempDataProvider.cs [hunk 4]
```diff
@@ -61,15 +63,22 @@ public class SessionStateTempDataProvider : ITempDataProvider
             // Accessing Session property will throw if the session middleware is not enabled.
             var session = context.Session;
 
-            var tempDataDictionary = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
+            Dictionary<string, object> tempDataDictionary = null;
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/SessionStateTempDataProvider.cs]: You probably don't have to initialize this.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/SessionStateTempDataProvider.cs [hunk 5]
```diff
@@ -153,7 +159,7 @@ public class SessionStateTempDataProvider : ITempDataProvider
                 session.Set(TempDataSessionStateKey, new byte[] { });
             }
 
-            return tempDataDictionary;
+            return tempDataDictionary ?? new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/SessionStateTempDataProvider.cs]: When would this be null?

Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/SessionStateTempDataProvider.cs]: This would be null when the `else` case runs.

**Comment by kichalla** [comment]
:shipit:

**Pull request merged**